hack/shoot-comparator/pkg/shoot/provider_matcher.go

```diff
@@ -77,6 +98,123 @@ func (m *ProviderMatcher) FailureMessage(_ interface{}) string {
 
 type propertyMatcher = struct {
 	types.GomegaMatcher
-	path     string
-	expected interface{}
+	path   string
+	actual interface{}
+}
+
+func idWorker(v interface{}) string {
+	if v == nil {
+		panic("nil value")
+	}
+
+	w, ok := v.(v1beta1.Worker)
+	if !ok {
+		panic("invalid type")
+	}
+
+	return w.Name
+}
+
+func workers(ws []v1beta1.Worker) gstruct.Elements {
```

Comment:
I think the matcher is a bit inconsistent.
The following fields are ignored:
- `kubernetes`
- `machine.architecture`
- `taints`
- `dataVolumes`
- `systemComponents`
- `machineControllerManagerSettings`
- `clusterAutoscaler`

On the other hand we take into comparison the following fieldsL
- `annotations`
- `caBundle`
- `cri`
- `labels`
- `sysctls`

As we discussed previously the `worker` field in Runtime spec directly uses `gardener.Worker` type. 

The following is also important:
- KEB sets only subset of worker fields:
  - `maxSurge`
  - `maxUnavailable`
  - `maximum`
  - `minimum`
  - `name`
  - `volume`
  - `zones`
  - `type`
- Converter sets only the following fields with the default values:
  - `machine.image.version`
  - `machine.image.name`   
- Migrator rewrites workers and sets `nil` for the following fields:
  - `machine.architecture`
  - `systemComponents`
  -  `cri`

I think we need to make it all consistent.
We can take one of the two approaches:
1. Migrator rewrites, and matcher compares only subset of workers fields.
2. Migrator rewrites full worker object, and matcher simply uses`beComparableTo`

The first option feels like complicating  a simple thing. It would allow to compare full object from Gardener with partial object from converter (so that I could compare shoot stored on KIM's volume with full object from Gardener). But if we compare potential gain with the cost it seems to not be worth it.
The second option seems to me to be more appealing due to the following reasons:
- Matchers code will be simpler, and unit tests shorter
- Scenario with comparing two objects stored by KIM/Provisioner will work without problems
-In case we compare object from converter with the one form Gardener the difference will be detected. But it will not be the case from migrator.   
- In the migration scenario we will be safe. As we rewrite whole workers object from the shoot, we will detect no changes. I assume Gardener won't update workers object after it was created.

Reply:
There are many ways how we can improve on this code.
I propose to do it in the different PR.

Reply:
For now we decided to explicitly ignore some fields from the comparison.